# **AI sem 4**
# **NLP project**
# **team members**


*   **Naman Agarwal**
*   **Mukund bihari gupta**

*   **Nishant Mohan**
*   **Prashant Kumar**

*   **Naman Katiyar**








# **GOAL**
# So our main goal is to develop a humanized essay/story generator that can take up a prompt and based on it generate a short story that resembles a human written story

# **STEP 1** -  Finding and processing the datsset


# choosing dataset wisely is of great importance when it comes to NLP tasks because the context or the emotions or the language our model uses are all dependent on the dataset.

# We are Using the writing propmts dataset from kaggle(link below)
https://www.kaggle.com/datasets/ratthachat/writing-prompts

# This dataset contains a huge corpus of human written essays over a wide variety of prompts it includes both formal and informal language and we found out to be suitable for our task

In [ ]:
import pandas as pd

In [ ]:
with open('../Dataset/writingPrompts/test.wp_source', 'r') as file:
    source_lines = file.readlines()
with open('../Dataset/writingPrompts/test.wp_target', 'r') as file:
    target_lines = file.readlines()

df_source = pd.DataFrame(source_lines, columns=['source'])
df_target = pd.DataFrame(target_lines, columns=['target'])

df_merged = pd.concat([df_source, df_target], axis=1)

df_merged.to_json('../PreProcessedDataset/test.json', orient='records', lines=True)

print(df_merged)

                                                  source  \
0      [ WP ] Leonardo DiCaprio in a fit of rage begi...   
1      [ CW ] Kill the writer in first-person narrati...   
2      [ EU ] Sean Bean has a hard time leaving his r...   
3      [ WP ] A kid doodling in a math class accident...   
4      [ WP ] As a Space marine you have an allowance...   
...                                                  ...   
15133  [ WP ] You are a devout christian , and you ju...   
15134  [ WP ] Describe a universe where a personal st...   
15135  [ WP ] Turning your girlfriend into a pumpkin ...   
15136  [ WP ] God answers every time you call His nam...   
15137  [ cw ] write about the strangest/scariest/sadd...   

                                                  target  
0      The wet marble floor pressed on his cheek like...  
1      It 's been three days since my boyfriend pisse...  
2      George fled . Not terrifically fast , not at h...  
3      It was dark and Levi was pretty sure

In [ ]:
df=pd.read_json('../PreProcessedDataset/test.json', lines=True)
df.head()

source  \
0  [ WP ] Leonardo DiCaprio in a fit of rage begi...   
1  [ CW ] Kill the writer in first-person narrati...   
2  [ EU ] Sean Bean has a hard time leaving his r...   
3  [ WP ] A kid doodling in a math class accident...   
4  [ WP ] As a Space marine you have an allowance...   

                                              target  
0  The wet marble floor pressed on his cheek like...  
1  It 's been three days since my boyfriend pisse...  
2  George fled . Not terrifically fast , not at h...  
3  It was dark and Levi was pretty sure he was ly...  
4  `` Hi dad ! '' <newline> <newline> The words s...

In [ ]:
with open('../Dataset/writingPrompts/train.wp_source', 'r') as file:
    source_lines = file.readlines()
with open('../Dataset/writingPrompts/train.wp_target', 'r') as file:
    target_lines = file.readlines()

df_source = pd.DataFrame(source_lines, columns=['source'])
df_target = pd.DataFrame(target_lines, columns=['target'])

df_merged = pd.concat([df_source, df_target], axis=1)

df_merged.to_json('../PreProcessedDataset/train.json', orient='records', lines=True)

In [ ]:
df=pd.read_json('../PreProcessedDataset/train.json', lines=True)
df.head()

source  \
0  [ WP ] You 've finally managed to discover the...   
1  [ WP ] The moon is actually a giant egg , and ...   
2  [ WP ] You find a rip in time walking through ...   
3  [ WP ] For years in your youth the same imagin...   
4  [ WP ] You glance at your watch 10:34 am , rou...   

                                              target  
0  So many times have I walked on ruins , the rem...  
1  -Week 18 aboard the Depth Reaver , Circa 2023-...  
2  I was feckin ' sloshed , mate . First time I e...  
3  “ No , no no no ... ” She backed up and turned...  
4  There 's a magical moment between wakefulness ...

In [ ]:
with open('../Dataset/writingPrompts/valid.wp_source', 'r') as file:
    source_lines = file.readlines()
with open('../Dataset/writingPrompts/valid.wp_target', 'r') as file:
    target_lines = file.readlines()

df_source = pd.DataFrame(source_lines, columns=['source'])
df_target = pd.DataFrame(target_lines, columns=['target'])

df_merged = pd.concat([df_source, df_target], axis=1)

df_merged.to_json('../PreProcessedDataset/valid.json', orient='records', lines=True)

In [ ]:
df=pd.read_json('../PreProcessedDataset/valid.json', lines=True)
df.head()

source  \
0  [ WP ] Every person in the world undergoes a `...   
1  [ WP ] Space mining is on the rise . The Space...   
2  [ WP ] `` I wo n't have time to explain all of...   
3  [ CW ] Write about a song . Each sentence must...   
4  [ EU ] You live in Skyrim . It is your job to ...   

                                              target  
0  Clancy Marguerian , 154 , private first class ...  
1  „… and the little duckling will never be able ...  
2  I wo n't have the time to explain all of this ...  
3  * '' [ Sally ] ( https : //www.youtube.com/wat...  
4  Light is a marvelous thing . It alone can turn...

# In the above code we have taken the dataset from kaggle and stored it in json format in our github and also on our respective drives for future use
**~by mukund**



---



# **Step 2** - Loading the stored dataset and making it suitable for training purpose

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_json('/content/drive/MyDrive/PreProcessedDataset/train.json',lines = True)
df

source  \
0       [ WP ] You 've finally managed to discover the...   
1       [ WP ] The moon is actually a giant egg , and ...   
2       [ WP ] You find a rip in time walking through ...   
3       [ WP ] For years in your youth the same imagin...   
4       [ WP ] You glance at your watch 10:34 am , rou...   
...                                                   ...   
272595  [ WP ] You wake up , extremely thirsty and dre...   
272596  [ WP ] After many years , you finally decide t...   
272597  [ WP ] In a world where people can only be kil...   
272598  [ WP ] Use a lyric from a song , or even the w...   
272599  [ CW ] [ PM ] Write your hero into a corner , ...   

                                                   target  
0       So many times have I walked on ruins , the rem...  
1       -Week 18 aboard the Depth Reaver , Circa 2023-...  
2       I was feckin ' sloshed , mate . First time I e...  
3       “ No , no no no ... ” She backed up and turned...  
4       There 's a magical moment between wakefulness ...  
...                                                   ...  
272595  `` He was a good man . '' <newline> <newline> ...  
272596  The Envelope <newline> <newline> It 's been 23...  
272597  17 people . Ive killed 17 people . I cant reme...  
272598  > So make the best of this test and do n't ask...  
272599  Bob dropped five of the Zeds , reloaded his Co...  

[272600 rows x 2 columns]



---



Here we are merging the prompt and the essay column because while generating the model should treat them as a single continuous entity because intutively what the model does is predict the next token based on the previous tokens so the llms like chatgpt or deepseek effectively are just continuing your prompt and the prompt has to be included in the context window



---



In [ ]:
df['merged'] = df['source'] + " " + df['target']
df

source  \
0       [ WP ] You 've finally managed to discover the...   
1       [ WP ] The moon is actually a giant egg , and ...   
2       [ WP ] You find a rip in time walking through ...   
3       [ WP ] For years in your youth the same imagin...   
4       [ WP ] You glance at your watch 10:34 am , rou...   
...                                                   ...   
272595  [ WP ] You wake up , extremely thirsty and dre...   
272596  [ WP ] After many years , you finally decide t...   
272597  [ WP ] In a world where people can only be kil...   
272598  [ WP ] Use a lyric from a song , or even the w...   
272599  [ CW ] [ PM ] Write your hero into a corner , ...   

                                                   target  \
0       So many times have I walked on ruins , the rem...   
1       -Week 18 aboard the Depth Reaver , Circa 2023-...   
2       I was feckin ' sloshed , mate . First time I e...   
3       “ No , no no no ... ” She backed up and turned...   
4       There 's a magical moment between wakefulness ...   
...                                                   ...   
272595  `` He was a good man . '' <newline> <newline> ...   
272596  The Envelope <newline> <newline> It 's been 23...   
272597  17 people . Ive killed 17 people . I cant reme...   
272598  > So make the best of this test and do n't ask...   
272599  Bob dropped five of the Zeds , reloaded his Co...   

                                                   merged  
0       [ WP ] You 've finally managed to discover the...  
1       [ WP ] The moon is actually a giant egg , and ...  
2       [ WP ] You find a rip in time walking through ...  
3       [ WP ] For years in your youth the same imagin...  
4       [ WP ] You glance at your watch 10:34 am , rou...  
...                                                   ...  
272595  [ WP ] You wake up , extremely thirsty and dre...  
272596  [ WP ] After many years , you finally decide t...  
272597  [ WP ] In a world where people can only be kil...  
272598  [ WP ] Use a lyric from a song , or even the w...  
272599  [ CW ] [ PM ] Write your hero into a corner , ...  

[272600 rows x 3 columns]

In [ ]:
special_char = ' # '
text = special_char.join(df['merged'].dropna().astype(str))
len(text)

944467148



---



Decreasing the size of our data because our session got crashed lol



---



In [ ]:
ok = int(0.01*len(text))
text = text[:ok]

In [ ]:
ok = int(0.01*len(text))
text = text[:ok]

In [ ]:
unique_chars = set(text)
print(unique_chars)

{'ê', 'F', 'o', '=', '!', 'è', 'р', 'ï', '،', '-', 'y', '6', 'ϐ', 'u', 'ل', '_', 'أ', '%', '\u200b', '0', 'á', 'B', ';', '´', 't', 'В', '⊙', 'ö', 'Q', 'ë', '－', 'ð', 'ß', 'щ', 'D', 'е', '≥', ' ', '^', '«', '(', 'V', '―', 'а', '.', '×', 'ا', 'ت', 'Z', 'آ', '?', 'ы', ')', 'м', 'م', '…', 'с', 'т', 'ю', 'Y', 'ó', 'k', 'Э', '№', 'd', 'L', 'П', '❧', 'ь', '–', '”', '*', 'ب', 'c', '|', '\x90', '$', 'a', 'о', '！', '’', 'r', 'n', 'T', 'R', 'ж', '4', 'E', 'm', '/', 'ջ', ',', '▦', 'О', 'Т', '9', '‽', '3', '\x10', 'I', 'A', 'ه', '~', '@', 'ę', 'ä', 'п', 'ط', 'ø', '▶', 'خ', 'W', 'Ֆ', '\\', '\x81', 'v', 'к', 'х', 'в', 'M', '®', 'س', '}', '5', '¡', 'g', '∞', '`', 'w', 'б', ']', '>', 'ذ', 'Ϡ', 'д', '{', 'л', 'ن', 'ك', '╬', '☂', 'Ö', 'f', '»', 'i', '2', 'q', 'ç', '£', '1', 'j', 'й', '◀', 'C', 'Δ', 'Д', 'e', 'h', '\ufeff', '+', '。', '&', 'Ԭ', 'P', '[', 'ف', 'و', '÷', 'ج', 'z', 'Н', 'b', 'É', '#', 'ч', 'p', 'N', '‘', '<', 'غ', 'С', 'X', '\n', 'H', '8', '，', 'x', 'º', '™', '“', 'J', 'ü', '7', 'ي', 'Л', 'Ч'



---



Above we can see that the dataset contains many letters not from english language (Honestly i dont know where these chinese alphabets came from lol so lets remove them using simple regex :)



---



In [ ]:
import re
allowed_chars = "a-zA-Z\n,\.! "  # allowed: english letters, newline, comma, period, exclamation mark, and space
allowed_chars_pattern = f"[^{allowed_chars}]"
cleaned_text = re.sub(allowed_chars_pattern, '', text)


In [ ]:
len(cleaned_text)

9087940

In [ ]:
unique_chars = set(cleaned_text)
print(unique_chars)

{'z', 'Y', 'F', 'v', 'b', 'k', 'd', 'o', 'p', 'N', 'M', '!', 'L', 'X', '\n', 'H', 'y', 'x', 'g', 'c', 'u', 'a', 'w', 'J', 'r', 'n', 's', 'T', 'R', 'B', 'f', 't', 'E', 'Q', 'm', 'i', 'q', 'D', ' ', ',', 'j', 'G', 'C', 'e', 'h', 'V', 'I', 'A', '.', 'S', 'U', 'Z', 'P', 'K', 'O', 'l', 'W'}


normal neural net


In [ ]:
chars = sorted(set(cleaned_text))
vocab_size = len(chars)
chars


['\n',
 ' ',
 '!',
 ',',
 '.',
 'A',
 'B',
 'C',
 'D',
 'E',
 'F',
 'G',
 'H',
 'I',
 'J',
 'K',
 'L',
 'M',
 'N',
 'O',
 'P',
 'Q',
 'R',
 'S',
 'T',
 'U',
 'V',
 'W',
 'X',
 'Y',
 'Z',
 'a',
 'b',
 'c',
 'd',
 'e',
 'f',
 'g',
 'h',
 'i',
 'j',
 'k',
 'l',
 'm',
 'n',
 'o',
 'p',
 'q',
 'r',
 's',
 't',
 'u',
 'v',
 'w',
 'x',
 'y',
 'z']



---



# Perfect!
# Now its necessary to note that which entities are being selected by us as tokens since we know the LLMs nowdays use probably some 4-5 letters together as one token but in our case for simplicity we consider a single letter as a token which will save time while training  

# Note - We may increase the length of tokens in future for better results

In [ ]:
letter_to_number = {letter: idx for idx, letter in enumerate(chars)}
encoded_text = [letter_to_number[char] for char in cleaned_text] # encoding the letters to numbers

better way to encode


In [ ]:
import torch

In [ ]:
stoi = { ch:i for i,ch in enumerate(chars) }
itos = { i:ch for i,ch in enumerate(chars) }
encode = lambda s: [stoi[c] for c in s] # encoder: take a string, output a list of integers
decode = lambda l: ''.join([itos[i] for i in l]) # decoder: take a list of integers, output a string


In [ ]:
data = torch.tensor(encode(cleaned_text), dtype=torch.long)
print(data.shape, data.dtype)
print(data[:1000])

torch.Size([9087940]) torch.int64
tensor([ 1, 27, 20,  1,  1, 29, 45, 51,  1, 52, 35,  1, 36, 39, 44, 31, 42, 42,
        55,  1, 43, 31, 44, 31, 37, 35, 34,  1, 50, 45,  1, 34, 39, 49, 33, 45,
        52, 35, 48,  1, 50, 38, 35,  1, 49, 35, 33, 48, 35, 50,  1, 50, 45,  1,
        39, 43, 43, 45, 48, 50, 31, 42, 39, 50, 55,  1,  4,  1, 23, 51, 34, 34,
        35, 44, 42, 55,  1,  3,  1,  8, 35, 31, 50, 38,  1, 31, 46, 46, 35, 31,
        48, 49,  1, 32, 35, 36, 45, 48, 35,  1, 55, 45, 51,  1,  3,  1, 38, 31,
        44, 34, 49,  1, 55, 45, 51,  1, 31,  1, 32, 51, 49, 39, 44, 35, 49, 49,
         1, 33, 31, 48, 34,  1,  3,  1, 31, 44, 34,  1, 49, 31, 55, 49,  1,  3,
         1,  1, 27, 38, 35, 44,  1, 55, 45, 51,  1, 48, 35, 31, 42, 39, 56, 35,
         1, 42, 39, 52, 39, 44, 37,  1, 36, 45, 48, 35, 52, 35, 48,  1, 49, 51,
        33, 41, 49,  1,  3,  1, 33, 31, 42, 42,  1, 50, 38, 39, 49,  1, 44, 51,
        43, 32, 35, 48,  1,  3,  1, 13,  1, 52, 35,  1, 37, 45, 50,  1, 31,  1,
      

In [ ]:
n = int(0.9*len(data)) # first 90% will be train, rest val
train_data = data[:n]
val_data = data[n:]

**credits~ Naman, Prashant**



---



# **NOW LETS START WITH THE MODELS**

# SO our main objective is not only to make a resume level project but to learn along the way. So we dont want to directly jump to final step which would be fine tuning a pretrained becuase that is as easy as watching a yotube video and copying the code from net

# Instead we want to familiarize ouselves with the history of machine learning how it started, what were the initial models that paved the way for the final boss i.e. transformers

# **LETS BEGIN!**

# **Model 1** - bigram model

# Overview of the Bigram Model in NLP
The bigram model is a type of statistical language model that estimates the probability of a word based solely on the immediately preceding word. This approach assumes that the likelihood of a word depends only on its direct predecessor, significantly simplifying the modeling of language compared to considering longer contexts. Despite this simplification, bigram models have proven useful in various natural language processing tasks such as speech recognition, text prediction, and machine translation.

# The First Research Paper on Bigram Models
One of the earliest and most influential works that laid the foundation for statistical language models—including bigram models—is Claude Shannon’s seminal paper "A Mathematical Theory of Communication" (1948)

# NOTE - below we have implemented a character level bigram model

In [ ]:
torch.manual_seed(1337)
batch_size = 6 # how many independent sequences will we process in parallel?
block_size = 16 # what is the maximum context length for predictions? will increase this later this is just for testing

def get_batch(split):
    # generate a small batch of data of inputs x and targets y
    data = train_data if split == 'train' else val_data
    ix = torch.randint(len(data) - block_size, (batch_size,))
    x = torch.stack([data[i:i+block_size] for i in ix])
    y = torch.stack([data[i+1:i+block_size+1] for i in ix])
    return x, y

xb, yb = get_batch('train')
print('inputs:')
print(xb.shape)
print(xb)
print('targets:')
print(yb.shape)
print(yb)

print('----')

for b in range(batch_size): # batch dimension
    for t in range(block_size): # time dimension
        context = xb[b, :t+1]
        target = yb[b,t]
        print(f"when input is {context.tolist()} the target: {target}")

inputs:
torch.Size([6, 16])
tensor([[39, 44, 37,  1, 39, 50,  1, 51, 44, 50, 39, 42,  1, 50, 38, 35],
        [55,  1,  4,  1,  1, 44, 35, 53, 42, 39, 44, 35,  1, 44, 35, 53],
        [50, 38, 35,  1, 49, 50, 48, 35, 35, 50, 49,  1, 50, 38, 35,  1],
        [55,  1,  4,  0,  1, 24, 38, 35,  1, 49, 46, 31, 48, 41, 42, 39],
        [46, 42, 39, 35, 49,  1, 53, 39, 50, 38,  1, 31, 44, 37, 35, 48],
        [34,  1, 33, 38, 35, 33, 41,  1, 36, 45, 48,  1, 55, 45, 51, 48]])
targets:
torch.Size([6, 16])
tensor([[44, 37,  1, 39, 50,  1, 51, 44, 50, 39, 42,  1, 50, 38, 35,  1],
        [ 1,  4,  1,  1, 44, 35, 53, 42, 39, 44, 35,  1, 44, 35, 53, 42],
        [38, 35,  1, 49, 50, 48, 35, 35, 50, 49,  1, 50, 38, 35,  1, 31],
        [ 1,  4,  0,  1, 24, 38, 35,  1, 49, 46, 31, 48, 41, 42, 39, 35],
        [42, 39, 35, 49,  1, 53, 39, 50, 38,  1, 31, 44, 37, 35, 48,  1],
        [ 1, 33, 38, 35, 33, 41,  1, 36, 45, 48,  1, 55, 45, 51, 48, 49]])
----
when input is [39] the target: 44
when input is 

In [ ]:
import torch
import torch.nn as nn
from torch.nn import functional as F
torch.manual_seed(1337)

vocab_size = len(chars)

class BigramLanguageModel(nn.Module):

    def __init__(self, vocab_size):
        super().__init__()
        # each token directly reads off the logits for the next token from a lookup table
        self.token_embedding_table = nn.Embedding(vocab_size, vocab_size)

    def forward(self, idx, targets=None):

        # idx and targets are both (B,T) tensor of integers
        logits = self.token_embedding_table(idx) # (B,T,C)

        if targets is None:
            loss = None
        else:
            B, T, C = logits.shape
            logits = logits.view(B*T, C)
            targets = targets.view(B*T)
            loss = F.cross_entropy(logits, targets)

        return logits, loss

    def generate(self, idx, max_new_tokens):
        # idx is (B, T) array of indices in the current context
        for _ in range(max_new_tokens):
            # get the predictions
            logits, loss = self(idx)
            # focus only on the last time step
            logits = logits[:, -1, :] # becomes (B, C)
            # apply softmax to get probabilities
            probs = F.softmax(logits, dim=-1) # (B, C)
            # sample from the distribution
            idx_next = torch.multinomial(probs, num_samples=1) # (B, 1)
            # append sampled index to the running sequence
            idx = torch.cat((idx, idx_next), dim=1) # (B, T+1)
        return idx

m = BigramLanguageModel(vocab_size)
logits, loss = m(xb, yb)
print(logits.shape)
print(loss)

print(decode(m.generate(idx = torch.zeros((1, 1), dtype=torch.long), max_new_tokens=100)[0].tolist()))


torch.Size([640, 57])
tensor(4.5551, grad_fn=<NllLossBackward0>)

Cbp,HZOccjxqerAiK
wwVpBxbhuBr rXvaPywWhA
nL
rbqFSiytRjATRHP.cTxDnYVSQXZ.Q!YwLjDPV

KBhmpNcpQChsCVXbX




---



# Aove is some random stuff generated by our intial logits before training



---



In [ ]:
optimizer = torch.optim.AdamW(m.parameters(), lr=1e-3)

In [ ]:
batch_size = 32
for steps in range(300):

    # sample a batch of data
    xb, yb = get_batch('train')

    # evaluate the loss
    logits, loss = m(xb, yb)
    optimizer.zero_grad(set_to_none=True)
    loss.backward()
    optimizer.step()

print(loss.item())


2.4821109771728516


In [ ]:
print(decode(m.generate(idx = torch.zeros((1, 1), dtype=torch.long), max_new_tokens=500)[0].tolist()))


 orTTho nXD wais h Uphaus t aGinewu s nUDf the bs toomonRS s aseXc!gFFUz sy gh  medsk Stpownd flyqu . trd Sh ben tperso cho tbjnthaneneryoaldeveL kindmZ
waLurighanouinewlal wlercke , , veIAee douxtlzalamed athit ly t pOnRes vesk winedoud onte,T tKane  tMSaHEMye mle n d I . angeNGm reeleRYomou waro in  tasEfat, hweay  ,oNNMcawligho thedrechevozyDQ, owoutXYOvene Haca homutXert ,FAsasit leXke ckictame G s h toKpus . stThont ththenYe semulippe boANshe ar g eng ai miors OlllisheaZanTOn sqaghapaHithon




---



# Above is the generation after training the bigram model
# Our loss is 2.48 here which still prety high if we talk about a character level model but and as you can see it still generates some giberish but the ouput is more structured and you can also see word formation that insertion of spaces which wasnt seen in the random stuff generated above. This is a hint of progress
# lets keep going

**credits - Naman and katta bhai**

# **Main problem** - the bigram model only looks at the previous charcter for prediction so lets give the other characters some attention

# Note - here we are trying to make the historic models better using the 2017 paper **Attention is all you need**

In [ ]:
import torch
import torch.nn as nn
from torch.nn import functional as F


In [ ]:
batch_size = 20 # how many independent sequences will we process in parallel?
block_size = 32 # what is the maximum context length for predictions?
max_iters = 5000
eval_interval = 100
learning_rate = 1e-3
device = 'cuda' if torch.cuda.is_available() else 'cpu'
eval_iters = 200
n_embd = 64
n_head = 4
n_layer = 4
dropout = 0.0

In [ ]:
def get_batch(split):
    # generate a small batch of data of inputs x and targets y
    data = train_data if split == 'train' else val_data
    ix = torch.randint(len(data) - block_size, (batch_size,))
    x = torch.stack([data[i:i+block_size] for i in ix])
    y = torch.stack([data[i+1:i+block_size+1] for i in ix])
    x, y = x.to(device), y.to(device)
    return x, y

@torch.no_grad()
def estimate_loss():
    out = {}
    model.eval()
    for split in ['train', 'val']:
        losses = torch.zeros(eval_iters)
        for k in range(eval_iters):
            X, Y = get_batch(split)
            logits, loss = model(X, Y)
            losses[k] = loss.item()
        out[split] = losses.mean()
    model.train()
    return out

class Head(nn.Module):
    """ one head of self-attention """

    def __init__(self, head_size):
        super().__init__()
        self.key = nn.Linear(n_embd, head_size, bias=False)
        self.query = nn.Linear(n_embd, head_size, bias=False)
        self.value = nn.Linear(n_embd, head_size, bias=False)
        self.register_buffer('tril', torch.tril(torch.ones(block_size, block_size)))

        self.dropout = nn.Dropout(dropout)

    def forward(self, x):
        B,T,C = x.shape
        k = self.key(x)   # (B,T,C)
        q = self.query(x) # (B,T,C)
        # compute attention scores ("affinities")
        wei = q @ k.transpose(-2,-1) * C**-0.5 # (B, T, C) @ (B, C, T) -> (B, T, T)
        wei = wei.masked_fill(self.tril[:T, :T] == 0, float('-inf')) # (B, T, T)
        wei = F.softmax(wei, dim=-1) # (B, T, T)
        wei = self.dropout(wei)
        # perform the weighted aggregation of the values
        v = self.value(x) # (B,T,C)
        out = wei @ v # (B, T, T) @ (B, T, C) -> (B, T, C)
        return out

class MultiHeadAttention(nn.Module):
    """ multiple heads of self-attention in parallel """

    def __init__(self, num_heads, head_size):
        super().__init__()
        self.heads = nn.ModuleList([Head(head_size) for _ in range(num_heads)])
        self.proj = nn.Linear(n_embd, n_embd)
        self.dropout = nn.Dropout(dropout)

    def forward(self, x):
        out = torch.cat([h(x) for h in self.heads], dim=-1)
        out = self.dropout(self.proj(out))
        return out

class FeedFoward(nn.Module):
    """ a simple linear layer followed by a non-linearity """

    def __init__(self, n_embd):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(n_embd, 4 * n_embd),
            nn.ReLU(),
            nn.Linear(4 * n_embd, n_embd),
            nn.Dropout(dropout),
        )

    def forward(self, x):
        return self.net(x)

class Block(nn.Module):
    """ Transformer block: communication followed by computation """

    def __init__(self, n_embd, n_head):
        # n_embd: embedding dimension, n_head: the number of heads we'd like
        super().__init__()
        head_size = n_embd // n_head
        self.sa = MultiHeadAttention(n_head, head_size)
        self.ffwd = FeedFoward(n_embd)
        self.ln1 = nn.LayerNorm(n_embd)
        self.ln2 = nn.LayerNorm(n_embd)

    def forward(self, x):
        x = x + self.sa(self.ln1(x))
        x = x + self.ffwd(self.ln2(x))
        return x

# super simple bigram model
class BigramLanguageModel(nn.Module):

    def __init__(self):
        super().__init__()
        # each token directly reads off the logits for the next token from a lookup table
        self.token_embedding_table = nn.Embedding(vocab_size, n_embd)
        self.position_embedding_table = nn.Embedding(block_size, n_embd)
        self.blocks = nn.Sequential(*[Block(n_embd, n_head=n_head) for _ in range(n_layer)])
        self.ln_f = nn.LayerNorm(n_embd) # final layer norm
        self.lm_head = nn.Linear(n_embd, vocab_size)

    def forward(self, idx, targets=None):
        B, T = idx.shape

        # idx and targets are both (B,T) tensor of integers
        tok_emb = self.token_embedding_table(idx) # (B,T,C)
        pos_emb = self.position_embedding_table(torch.arange(T, device=device)) # (T,C)
        x = tok_emb + pos_emb # (B,T,C)
        x = self.blocks(x) # (B,T,C)
        x = self.ln_f(x) # (B,T,C)
        logits = self.lm_head(x) # (B,T,vocab_size)

        if targets is None:
            loss = None
        else:
            B, T, C = logits.shape
            logits = logits.view(B*T, C)
            targets = targets.view(B*T)
            loss = F.cross_entropy(logits, targets)

        return logits, loss

    def generate(self, idx, max_new_tokens):
        # idx is (B, T) array of indices in the current context
        for _ in range(max_new_tokens):
            # crop idx to the last block_size tokens
            idx_cond = idx[:, -block_size:]
            # get the predictions
            logits, loss = self(idx_cond)
            # focus only on the last time step
            logits = logits[:, -1, :] # becomes (B, C)
            # apply softmax to get probabilities
            probs = F.softmax(logits, dim=-1) # (B, C)
            # sample from the distribution
            idx_next = torch.multinomial(probs, num_samples=1) # (B, 1)
            # append sampled index to the running sequence
            idx = torch.cat((idx, idx_next), dim=1) # (B, T+1)
        return idx

model = BigramLanguageModel()
m = model.to(device)
# print the number of parameters in the model
print(sum(p.numel() for p in m.parameters())/1e6, 'M parameters')

# create a PyTorch optimizer
optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)

for iter in range(max_iters):

    # every once in a while evaluate the loss on train and val sets
    if iter % eval_interval == 0 or iter == max_iters - 1:
        losses = estimate_loss()
        print(f"step {iter}: train loss {losses['train']:.4f}, val loss {losses['val']:.4f}")

    # sample a batch of data
    xb, yb = get_batch('train')

    # evaluate the loss
    logits, loss = model(xb, yb)
    optimizer.zero_grad(set_to_none=True)
    loss.backward()
    optimizer.step()

# generate from the model
context = torch.zeros((1, 1), dtype=torch.long, device=device)
print(decode(m.generate(context, max_new_tokens=2000)[0].tolist()))


0.208697 M parameters
step 0: train loss 4.2872, val loss 4.2902
step 100: train loss 2.4598, val loss 2.4526
step 200: train loss 2.3310, val loss 2.3426
step 300: train loss 2.2324, val loss 2.2336
step 400: train loss 2.1669, val loss 2.1581
step 500: train loss 2.0889, val loss 2.0983
step 600: train loss 2.0325, val loss 2.0257
step 700: train loss 1.9712, val loss 1.9892
step 800: train loss 1.9455, val loss 1.9462
step 900: train loss 1.9084, val loss 1.9122
step 1000: train loss 1.8866, val loss 1.8823
step 1100: train loss 1.8791, val loss 1.8564
step 1200: train loss 1.8408, val loss 1.8550
step 1300: train loss 1.8172, val loss 1.8178
step 1400: train loss 1.8004, val loss 1.8152
step 1500: train loss 1.7793, val loss 1.7898
step 1600: train loss 1.7601, val loss 1.7778
step 1700: train loss 1.7587, val loss 1.7673
step 1800: train loss 1.7510, val loss 1.7493
step 1900: train loss 1.7406, val loss 1.7478
step 2000: train loss 1.7127, val loss 1.7279
step 2100: train loss 1.

# Interesting, So we tried adding a multi attention head to our charcter level model and the results didnt disappoint
# the loss has went down to 1.5 which is prety good also there is negligible divergence between train and val loss which is a good sign and shows there is no overfitting

# Speaking of the ouput we can observe drastic improvement in terms that the text now resembles english language we can observe some legitimate words and in comparison to normal bigram it is very much improved

# **Problem** - The bottleneck now is the token length so no matter how much we train the charcter level model it is quite difficult for it to generate a complete essay from letter level predictions
# So Next step would be Switching to words and hoping for better results

**Credits - Naman,Nishant**



---





---



# **Model 2** - Multi Layer Perceptron (using words as tokens)